service/src/main/java/com/codeforcommunity/processor/ReservationProcessorImpl.java

```diff
@@ -0,0 +1,292 @@
+package com.codeforcommunity.processor;
+
+import static org.jooq.generated.Tables.*;
+
+import com.codeforcommunity.api.IReservationProcessor;
+import com.codeforcommunity.auth.JWTData;
+import com.codeforcommunity.dto.reservation.*;
+import com.codeforcommunity.enums.PrivilegeLevel;
+import com.codeforcommunity.enums.ReservationAction;
+import com.codeforcommunity.enums.TeamRole;
+import com.codeforcommunity.exceptions.*;
+import java.sql.Timestamp;
+import java.util.Optional;
+import org.jooq.DSLContext;
+import org.jooq.generated.tables.records.*;
+
+public class ReservationProcessorImpl implements IReservationProcessor {
+
+  private final DSLContext db;
+
+  public ReservationProcessorImpl(DSLContext db) {
+    this.db = db;
+  }
+
+  public Optional<ReservationsRecord> lastAction(int blockId) {
+    return Optional.ofNullable(
+        db.selectFrom(RESERVATIONS)
+            .where(RESERVATIONS.BLOCK_ID.eq(blockId))
+            .orderBy(RESERVATIONS.PERFORMED_AT.desc())
+            .limit(1)
+            .fetchOne());
+  }
+
+  /**
+   * Checks if the user is not NONE or PENDING, and therefore on the team
+   *
+   * @param userId the id of the user
+   * @param teamId the id of the team
+   * @return a boolean indicating if they're on the team
+   */
+  private boolean isOnTeam(int userId, int teamId) {
+    if (!db.fetchExists(
+            db.selectFrom(USERS_TEAMS)
+                    .where(USERS_TEAMS.USER_ID.eq(userId))
+                    .and(USERS_TEAMS.TEAM_ID.eq(teamId))
+                    .and(USERS_TEAMS.TEAM_ROLE.notEqual(TeamRole.None))
+                    .and(USERS_TEAMS.TEAM_ROLE.notEqual(TeamRole.PENDING)))) {
+      throw new UserNotOnTeamException(userId, teamId);
+    }
+    return true;
+  }
+
+  private void basicChecks(int blockId, Integer userId, Integer teamId) {
+    if (!db.fetchExists(db.selectFrom(BLOCKS).where(BLOCKS.ID.eq(blockId)))) {
+      throw new ResourceDoesNotExistException(blockId, "block");
+    }
+
+    if (userId != null && !db.fetchExists(db.selectFrom(USERS).where(USERS.ID.eq(userId)))) {
+        throw new UserDoesNotExistException(userId);
+    }
+
+    if (teamId != null && !db.fetchExists(db.selectFrom(TEAMS).where(TEAMS.ID.eq(teamId)))) {
+        throw new ResourceDoesNotExistException(teamId, "team");
+    }
+
+    if (userId != null && teamId != null) {
```

Comment:
If we change isOnTeam to return true or false then you can just add it to the end of this if clause and throw the exception inside the if